tools/loader/Loader.cpp

```diff
@@ -366,7 +370,26 @@ void Loader::runInference(PlaceholderBindings &bindings, size_t batchSize) {
     timer.startTimer();
   }
   for (unsigned i = 0; i < iterationsOpt; i++) {
-    EE_.run(bindings);
+    std::unique_ptr<PlaceholderBindings> phBindings(&bindings);
+    std::unique_ptr<ExecutionContext> context =
+        llvm::make_unique<ExecutionContext>(std::move(phBindings));
+    std::promise<void> runPromise;
+    auto fut = runPromise.get_future();
+    llvm::Error runErr = llvm::Error::success();
+    hostManager_->runNetwork(
+        modelPathOpt[0], std::move(context),
+        [&runPromise, &runErr](runtime::RunIdentifierTy, llvm::Error err,
+                               std::unique_ptr<ExecutionContext> contextPtr) {
+          // Don't delete ph bindings.
```

Comment:
So, maybe this is out of scope for this precise diff, but the dance of creating a unique_ptr<ExecutionContext>, std::move-ing it in to a lambda, then std::move-ing out the unique_ptr<PlaceholderBindings> looks really sketchy.  Can that be avoided?

Reply:
There are two things interacting here:
* The Runtime interface is a unique_ptr because explicit ownership transfers reduce the incidence of a class of production bugs. We can always make writing the code easier to write by switching back to raw pointer, and so long as someone else is going to do all the object lifetime bug debugging i'm fine with that :)
* This code creates the PlaceholderBindings on the stack, which requires the hack below (because despite being in the unique_ptr the code below doesn't really own the PlaceholderBindings and we have to make sure it doesn't get deleted). If we created the PlaceholderBindings on the heap, or always as part of the ExecutionContext then it's not so big a deal. Alternatively if we had a better continuation abstraction (such as folly::Future) we wouldn't need the callbacks, but in the version of standard C++ we support this was what we have to work with.

Reply:
I guess I'm fine with the API taking a unique_ptr since runNetwork is intended to "consume" the bindings, and my objection is really to the second part, which is the weird move-out-and-release pattern that prevents the consumption from happening.

Maybe we should be copying the bindings on each iteration?  That adds a bit of overhead but is possibly more realistic anyways since usually a network run will need to make new bindings anyways.

Reply:
Why would we need new bindings? We could load new inputs into the existing input tensors by using DMA or RDMA, and in that case reuse the bindings forever. I think that it makes more sense to keep the bindings in a place that ensures the length of their lifetime and just pass them as pointer.

Reply:
@nadavrot I agree with your conclusion that we should keep the bindings in a place that ensures the length of lifetime.  The current code that passes unique_ptrs around and strategically releases them appears to be both more verbose and more error-prone.

But to your point of why we need new bindings: on the input side, we currently bind tensors directly to whatever memory we're given (e.g., https://github.com/pytorch/glow/blob/master/lib/Onnxifi/Base.cpp#L132) and rely on the device-specific code to manage DMA buffer re-use (if needed/possible).  But that means a call to runNetwork needs updated bindings since the caller doesn't guarantee where the tensors are in host memory.

Reply:
@bertmaher Thank you for explaining this to me. Makes sense.

Reply:
> The current code that passes unique_ptrs around and strategically releases them appears to be both more verbose and more error-prone.

Yes, agree this is bad - but it's not an intrinsic issue of the interface, its because the loader and the runtime disagree here. We should change the loader.